src/three_ary_cuckoo.rs

```diff
@@ -0,0 +1,235 @@
+            buckets: vec![None; capacity],
+            bucket_size: capacity / 3,
+            meta: MetaMap::new(capacity, meta_bits),
+            len: 0,
+        }
+    }
+
+    // (hash, [bucket_a, bucket_b, bucket_c])
+    fn buckets(&self, key: u64) -> (u64, [usize; 3]) {
+        let hash_a = self.hashers[0].hash_one(key);
+        let bucket_a = (hash_a % self.bucket_size as u64) as usize;
+        let mut bucket_b = (self.hashers[1].hash_one(key) % self.bucket_size as u64) as usize;
+        let mut bucket_c = (self.hashers[2].hash_one(key) % self.bucket_size as u64) as usize;
+
+        // each bucket has the same size, `self.bucket_size`,
+        // but they live in the same array just with different offsets
+        bucket_b += self.bucket_size;
+        bucket_c += 2 * self.bucket_size;
+
+        (hash_a, [bucket_a, bucket_b, bucket_c])
+    }
+
+    fn set_bucket(&mut self, bucket: usize, key: u64, hash: u64) {
+        self.buckets[bucket] = Some(key);
+        self.meta.set_full(bucket, Metadata::Hash(hash));
+    }
+
+    fn clear_bucket(&mut self, bucket: usize) {
+        self.buckets[bucket] = None;
+        self.meta.set_empty(bucket);
+    }
+}
+
+impl Map for ThreeAryCuckoo {
+    fn len(&self) -> usize {
+        self.len
+    }
+
+    fn capacity(&self) -> usize {
+        self.buckets.len()
+    }
+
+    fn probe(&self, key: u64) -> Probe {
+        let (hash, [bucket_a, bucket_b, bucket_c]) = self.buckets(key);
+
+        let mut probes = 0;
+
+        if !self.meta.hint_not_match(bucket_a, hash) {
+            probes += 1;
+            if self.buckets[bucket_a] == Some(key) {
+                return Probe {
+                    contained: true,
+                    probes,
+                };
+            }
+        }
+
+        if !self.meta.hint_not_match(bucket_b, hash) {
+            probes += 1;
+            if self.buckets[bucket_b] == Some(key) {
+                return Probe {
+                    contained: true,
+                    probes,
+                };
+            }
+        }
+
+        if !self.meta.hint_not_match(bucket_c, hash) {
+            probes += 1;
+            if self.buckets[bucket_c] == Some(key) {
+                return Probe {
+                    contained: true,
+                    probes,
+                };
+            }
+        }
+
+        Probe {
+            contained: false,
+            probes,
+        }
+    }
+
+    fn insert(&mut self, key: u64) -> Update {
+        const MAX_CHAIN: usize = 128;
+
+        let mut update = Update {
+            total_probes: 0,
+            total_writes: 1,
+            completed: true,
+        };
+
+        let mut active_key = key;
+        let mut key_info = self.buckets(key);
+
+        // test for presence.
+        {
+            let (hash, [bucket_a, bucket_b, bucket_c]) = key_info;
+
+            // TODO: not sure why in cuckoo.rs the presence check
+            // is applied only to bucket_b and not bucket_a
```

Comment:
it is applied only to B because it is applied to A directly afterwards in the first iteration of the loop. otherwise you can double-count the probe